File: problem_022.rs
// -*- compile-command: "rustc -o problem_022_rs problem_022.rs" -*-
// Copyright (c) 2016 Michael Caldwell
use std::fs::File;
use std::io::Read;

fn name_score(name: String) -> u64 {
    let mut score = 0;
    for c in name.chars() {
        if c == '"' { continue; }
        score += c as u64 - '@' as u64;
    }
    return score;
}

fn main() {
    println!("Project Euler - Problem 22:");
    println!("Sort 5000 names and calculate the sum of their name scores.\n");

    // Import data from file
    let mut data = String::new();
    let mut f = File::open("problem_022.txt").unwrap();
    f.read_to_string(&mut data).unwrap();

    // Split names into vector
    let mut names: Vec<&str> = data.split(',').collect();

    // Sort names alphabetically
    names.sort();

    // Calculate name score
    let mut total = 0;
    let mut index = 1;
    for name in names {
        let score = name_score(name.to_string());
        total += score * index;
        index += 1;
    }

    println!("{}", total);
}

File: problem_022.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "swiftc problem_022.swift -o problem_022" -*-
// Copyright (c) 2016 Michael Caldwell
import Cocoa
import Darwin
import Foundation

print("Project Euler - Problem 22:")
print("Sort 5000 names and calculate the sum of their name scores.\n")

func name_score(name: String) -> Int {
    var score = 0
    let characterArray = Array(name.characters)
    for character in characterArray {
        // Convert to ASCII value
        let scaler = String(character).unicodeScalars
        let asciiValue = scaler[scaler.startIndex].value
        // Skip quotes
        if character == "\"" { continue }
        // Convert to alphabetic value
        let charValue = asciiValue - UInt32("@")
        score += Int(charValue)
    }
    return score
}

// Import data from file
let file: FileHandle? = FileHandle(forReadingAtPath: "problem_022.txt")
if file == nil {
    print("Unable to open file.")
    exit(1)
}

// Read all the data
let data = file?.readDataToEndOfFile()

// Close the file
file?.closeFile()

// Convert our data to string
let dataAsString = String(data: data!, encoding: String.Encoding.utf8)

// Separate names into array
var names = dataAsString!._split(separator: ",")

// Sort names alphabetically
names.sort() { return $0 < $1 }

// Calculate name score
var total = 0
for i in 1...names.count {
    let score = name_score(name: names[i-1])
    total += score * i
}

print(total)
